### ISSUE TITLE
Fakewifi dont solve connection problem for some apps

### ISSUE THREAD

**Issue by ruditimmermans** [waydroid/waydroid#619]
We have here in Belgium and Bank app called Argenta, but it seems the change something into the app as it was working before and you can not install older version when then you get a message you need to update to lasted version without an update the app do not works.

So it seems it need a real wifi connection, i try with fakewifi but i still get no internet connection, so i think for more people into europa will get this problem into some time, Waydroid need a real Wifi connection like Alien Dalvik has on Sailfish OS there the app works without any problems.

Here is a link into the Play store:

https://play.google.com/store/apps/details?id=be.argenta.bankieren&hl=en

You can install the app for testing you dont need any account you get directly the message when you start the app there is no internet connection....

**Comment by NicolasDerumigny** [comment]
I don't know whether the issue is the same or not, but I encounter the same problem with "Moto Journal" application

https://play.google.com/store/apps/details?id=com.presstalis.motojournal&gl=US

I can log in, restore the purchases but not download issues (it's a magazine app) nor do anything network-related from the application (get Facebook page, etc) apart login. Logcat show:

```txt
12-27 13:03:18.504   278   356 D UntrustedWifiNetworkFactory: got request NetworkRequest [ TRACK_DEFAULT id=31, [ Capabilities: INTERNET&NOT_RESTRICTED&TRUSTED Uid: 10163 AdministratorUids: [] RequestorUid: 10163 RequestorPackageName: com.presstalis.motojournal] ] with score 70 and providerId 1
```

which may explain some things?

Other bug: the application crashs at startup as it seems not to be able to access its `cache` directory (`/media/0/Android/data/com,pesstalis.motojournal/cache`) that needs to be created and `chown`-ed by hand. After that, the application can open.

**Comment by NicolasDerumigny** [comment]
After further digging, the `mac80211_hwsim` kernel modul can be used to simulate a WiFi network connection. This is apparently used in Google's Android Emulator. This may be one path worth exploring to better handle network (as suggested by #251 ).

**Comment by ruditimmermans** [comment]
Oh this sound like good news ! thanks Nicolas for looking on this let's hope the Waydroid devs would fix it ;)

**Comment by NicolasDerumigny** [comment]
I have dig a bit more. Setting up a virtual wifi interface is easy on the host:
`modprobe mac80211_hwsim && hostapd hostapd.conf` with [this config](https://chromium.googlesource.com/chromiumos/third_party/hostap/+/0.12.362.B/mac80211_hwsim/hostapd.conf) to set up wlan0 as an access point.
Then, wlan1 can be passthrough the container with `lxc-device -P /var/lib/waydroid/lxc -n waydroid -- add wlan1`
However, the WiFi still cannot be turned on from WayDroid, `waydroid logcat` shows:
```log
12-29 11:16:26.669   262   279 I EthernetTracker: interfaceLinkStateChanged, iface: wlan1, up: false
12-29 11:16:26.673   262  1292 I WifiService: updateInterfaceIpState uid=1073
12-29 11:16:26.673   262   336 D WifiService: updateInterfaceIpState: ifaceName=wlan1 mode=-1 previous LOHS mode= -1
12-29 11:16:26.681   262   332 E NetworkStats: problem reading network stats
12-29 11:16:26.681   262   332 E NetworkStats: java.lang.IllegalStateException: problem parsing tethering stats: 
12-29 11:16:26.681   262   332 E NetworkStats: 	at com.android.server.NetworkManagementService$NetdTetheringStatsProvider.getTetherStats(NetworkManagementService.java:1714)
12-29 11:16:26.681   262   332 E NetworkStats: 	at com.android.server.NetworkManagementService.getNetworkStatsTethering(NetworkManagementService.java:1754)
12-29 11:16:26.681   262   332 E NetworkStats: 	at com.android.server.net.NetworkStatsService.getNetworkStatsTethering(NetworkStatsService.java:1905)
12-29 11:16:26.681   262   332 E NetworkStats: 	at com.android.server.net.NetworkStatsService.getNetworkStatsUidDetail(NetworkStatsService.java:1883)
12-29 11:16:26.681   262   332 E NetworkStats: 	at com.android.server.net.NetworkStatsService.recordSnapshotLocked(NetworkStatsService.java:1389)
12-29 11:16:26.681   262   332 E NetworkStats: 	at com.android.server.net.NetworkStatsService.performPollLocked(NetworkStatsService.java:1508)
12-29 11:16:26.681   262   332 E NetworkStats: 	at com.android.server.net.NetworkStatsService.performPoll(NetworkStatsService.java:1459)
12-29 11:16:26.681   262   332 E NetworkStats: 	at com.android.server.net.NetworkStatsService.access$100(NetworkStatsService.java:171)
12-29 11:16:26.681   262   332 E NetworkStats: 	at com.android.server.net.NetworkStatsService$2.onReceive(NetworkStatsService.java:1132)
12-29 11:16:26.681   262   332 E NetworkStats: 	at android.app.LoadedApk$ReceiverDispatcher$Args.lambda$getRunnable$0$LoadedApk$ReceiverDispatcher$Args(LoadedApk.java:1580)
12-29 11:16:26.681   262   332 E NetworkStats: 	at android.app.-$$Lambda$LoadedApk$ReceiverDispatcher$Args$_BumDX2UKsnxLVrE6UJsJZkotuA.run(Unknown Source:2)
12-29 11:16:26.681   262   332 E NetworkStats: 	at android.os.Handler.handleCallback(Handler.java:938)
12-29 11:16:26.681   262   332 E NetworkStats: 	at android.os.Handler.dispatchMessage(Handler.java:99)
12-29 11:16:26.681   262   332 E NetworkStats: 	at android.os.Looper.loop(Looper.java:223)
12-29 11:16:26.681   262   332 E NetworkStats: 	at android.os.HandlerThread.run(HandlerThread.java:67)
12-29 11:16:26.681   262   332 E NetworkStats: Caused by: android.os.ServiceSpecificException: [Remote I/O error] : failed to parse IPv4 tether stats:
12-29 11:16:26.681   262   332 E NetworkStats:  (code 121)
12-29 11:16:26.681   262   332 E NetworkStats: 	at android.os.Parcel.createExceptionOrNull(Parcel.java:2388)
12-29 11:16:26.681   262   332 E NetworkStats: 	at android.os.Parcel.createException(Parcel.java:2358)
12-29 11:16:26.681   262   332 E NetworkStats: 	at android.os.Parcel.readException(Parcel.java:2341)
12-29 11:16:26.681   262   332 E NetworkStats: 	at android.os.Parcel.readException(Parcel.java:2283)
12-29 11:16:26.681   262   332 E NetworkStats: 	at android.net.INetd$Stub$Proxy.tetherGetStats(INetd.java:2015)
12-29 11:16:26.681   262   332 E NetworkStats: 	at com.android.server.NetworkManagementService$NetdTetheringStatsProvider.getTetherStats(NetworkManagementService.java:1712)
12-29 11:16:26.681   262   332 E NetworkStats: 	... 14 more
```
I assume the Android guest system cannot use the virtual device (lack of drivers?). I may continue digging, but I do not have a lot of time.
Note that `wlan1` is detected by Android once passed through:
```sh
# ip addr
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: eth0@if4: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
    link/ether 00:16:3e:f9:d3:03 brd ff:ff:ff:ff:ff:ff link-netnsid 0
    inet 192.168.240.112/24 brd 192.168.240.255 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 fe80::ab17:eb54:40ab:dba5/64 scope link stable-privacy 
       valid_lft forever preferred_lft forever
6: wlan1: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN group default qlen 1000
    link/ether 5e:62:b4:84:a9:37 brd ff:ff:ff:ff:ff:ff
```

However, I cannot connect with `wpa_supplicant`, because it is not present in the (GAPPS) ROM.

**Comment by aruraune** [comment]
I have this same issue with a Brazilian Bank app.

**Comment by ruditimmermans** [comment]
Update more apps seems not to works anymore on Waydroid the request real wifi or data....

https://play.google.com/store/apps/details?id=be.bmid.itsme&hl=en&gl=US&pli=1

https://play.google.com/store/apps/details?id=mobi.inthepocket.bcmc.bancontact&hl=en&gl=US

**Comment by dmarrazzo** [comment]
I'm experiencing a similar problem with google translate app:
![image](https://user-images.githubusercontent.com/2084600/214490045-3c50d5f6-ba83-4e22-9593-bade91af4294.png)

**Comment by ruditimmermans** [comment]
I think we need urgent a wifi connection i have add last 5 apps that check on this and need this... and i think more and more apps one work anymore on Waydroid so it would become useless when there is no real wifi after some time (;

**Comment by CreativeKittens** [comment]
Yeah same problem here, i think some app only checking if the system connecting to the internet or not wheter through wifi or data cellular,  if not the they throw an exception or some stuff

**Comment by ruditimmermans** [comment]
I quit using Waydroid and all this alternatieve Oses to many apps give troble and i'm tryed of that.

**Comment by CreativeKittens** [comment]
> I quit using Waydroid and all this alternatieve Oses to many apps give troble and i'm tryed of that.

do you have alternative way to run android app beside anbox and waydroid?

**Comment by ruditimmermans** [comment]
Any news on this or a fix for the real wifi on Ubuntu Touch....?

**Comment by coolst3r** [comment]
why dont we just bridge the internet via virtual network

**Comment by ruditimmermans** [comment]
> why dont we just bridge the internet via virtual network

I dont know if this would help for example apps here in Belgium the do checks if there is a real wifi connection....

**Comment by NicolasDerumigny** [comment]
> why dont we just bridge the internet via virtual network

That is exactly what is done, but the network is passed through an ethernet adapter and not a wireless. If you are talking about a build-in VPN interface on android, we can try but I am not optimistic as this will not enable wifi / data connection for apps that are checking specifically for that.

**Comment by backendrulz** [comment]
Any solution? The command waydroid prop set persist.waydroid.fake_wifi doesn't work either. Maybe a module for Xposed/LSPosed?

**Comment by NicolasDerumigny** [comment]
There will not be any solution unless the project integrates drivers for a virtual wifi interface.

**Comment by msulemvn** [comment]
To use Google Translate offline, here's a temporary fix:

1. Download the language you need.
2. DNS block the Translate API:

```bash
sudo waydroid shell
mount -o rw,remount /
nano /etc/hosts
```
Add the following line to /etc/hosts:
```
0.0.0.0		translate-pa.googleapis.com
```

**Comment by baptx** [comment]
> There will not be any solution unless the project integrates drivers for a virtual wifi interface.

@NicolasDerumigny I remember Anbox was connected to a virtual WiFi network called VirtWifi. It would be good to do the same with Waydroid to fix the Internet issue on some apps.

**Comment by baptx** [comment]
> Any solution? The command `waydroid prop set persist.waydroid.fake_wifi` doesn't work either. Maybe a module for Xposed/LSPosed?

@backendrulz The command worked for me, you have to use it like mentioned in a comment from the user Neptaz on https://www.reddit.com/r/waydroid/comments/1fjfdkd/wifi_in_waydroid_not_turning_on/ (replace package_name with your package name):
`waydroid prop set persist.waydroid.fake_wifi package_name`

The user said we need to restart the Waydroid session but I did not need to do it.
I found the package name with the command `ps -ef` in `adb shell` while running the app.

This is a good workaround but it would be more convenient to be able to be connected to a virtual WiFi network like done on Anbox with VirtWifi.

Update: after reading the documentation on https://docs.waydro.id/usage/waydroid-prop-options I was able to match all packages with this command:
`waydroid prop set persist.waydroid.fake_wifi "*"`
To revert changes, use this command:
`waydroid prop set persist.waydroid.fake_wifi ""`

Using fake_wifi helped me to fix apps crashing on Waydroid with these errors in `adb logcat`:
```
E AndroidRuntime: java.lang.RuntimeException: Error receiving broadcast Intent { act=android.net.conn.CONNECTIVITY_CHANGE flg=0x4200010 (has extras) }
E AndroidRuntime: Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'boolean java.lang.String.isEmpty()' on a null object reference
```

**Comment by NicolasDerumigny** [comment]
> @NicolasDerumigny I remember Anbox was connected to a virtual WiFi network called VirtWifi. It would be good to do the same with Waydroid to fix the Internet issue on some apps.

From what I see, this is a virtualbox mechanism to emulate a WiFi card. That is also what `mac80211_hwsim` is doing, but Android needs to know what to do with this device, and (as far as I tried several years ago) it did not.

I think `fake_wifi` works for applicatiosn that checks that a wifi connection *exists*, but some apps actually try to use the connection directly instead of using some generic interface (as the generic interface would, in fact, route through the virtual wired interface), and there it breaks.

**Comment by sahib453** [comment]
> `waydroid prop set persist.waydroid.fake_wifi "*"`

works for me, thanks

**Comment by Alexandr71001** [comment]
Any solutions? Fake wifi doesn't work for me(using wlan0)

**Comment by sahib453** [comment]
> Any solutions? Fake wifi doesn't work for me(using wlan0)

sorry but I have shifted to macOS now... and I don't remember how I made the fake wifi work

**Comment by ayasa520** [comment]
I found some code that uses `mac80211_hwsim` to implement virtual wifi for https://github.com/remote-android/. But my attempts in waydroid failed. 
https://github.com/TeddyNight/vendor_redroid/commit/673adb0388e9d35a52014ab5ffed16ff1ede99d6
https://github.com/TeddyNight/device_generic_goldfish/commit/5bcc9e2feaad9ebe16dfd13aca69f2c85323c18f

**Comment by yjmantilla** [comment]
If someone is trying the fakewifi with whatsapp, it worked for me, but im unsure how. I tried both fakewifi on com.whatsapp and on "*", waited quite a bit, didnt work.

Kept trying out of frustration, restarting the session, multiple times.

It suddenly worked, no idea what changed.

**Comment by ayasa520** [comment]
> If someone is trying the fakewifi with whatsapp, it worked for me, but im unsure how. I tried both fakewifi on com.whatsapp and on "*", waited quite a bit, didnt work.
> 
> Kept trying out of frustration, restarting the session, multiple times.
> 
> It suddenly worked, no idea what changed.

fake wifi has been removed https://github.com/waydroid/waydroid/issues/1514#issuecomment-3134033100

**Comment by baptx** [comment]
> I found some code that uses `mac80211_hwsim` to implement virtual wifi for https://github.com/remote-android/. But my attempts in waydroid failed. [TeddyNight/vendor_redroid@673adb0](https://github.com/TeddyNight/vendor_redroid/commit/673adb0388e9d35a52014ab5ffed16ff1ede99d6) [TeddyNight/device_generic_goldfish@5bcc9e2](https://github.com/TeddyNight/device_generic_goldfish/commit/5bcc9e2feaad9ebe16dfd13aca69f2c85323c18f)

There is a related issue here: https://github.com/waydroid/waydroid/issues/1494

Some Android applications cannot be used on Waydroid because a network error is displayed, even if it works fine for example using Android-x86 on virt-manager.
I found a solution by using Frida and this script:
https://codeshare.frida.re/@zionspike/bypass-wi-fi-check-on-android/

On https://stackoverflow.com/questions/57277759/getactivenetworkinfo-is-deprecated-in-api-29, I read the function getActiveNetworkInfo used by the script is deprecated since API 29 (Android 10) but on Android 13, it used the second part of the Frida script, using NetworkCapabilities.
To make the script work, I had to call `bypass_wifi_check();` at the end of the file or remove `function bypass_wifi_check() {...}` and move the code inside:
```
Java.perform(function () {
...
});
```
After downloading the latest frida-server android-x86_64 version from https://github.com/frida/frida/releases, I ran these commands:
```
adb push frida-server /data/local/tmp
sudo waydroid shell
```

You need to make the file executable the first time you run it:
`chmod +x /data/local/tmp/frida-server`

The server can now be started with this command:
`/data/local/tmp/frida-server`

It should then be possible to execute the network bypass script with this command:
`frida -U -f APPLICATION_ID -l wifi_check_bypass.js`

However there is currently a bug that needs a workaround with a script that I had to customize:
https://github.com/frida/frida/issues/3605#issuecomment-4060657948

I successfully bypassed the network check of an Android application, using Frida version 17.8.1 on my computer and Waydroid with Android 13.
It is also working with the latest Frida version 17.8.2 released today.

**Comment by ayasa520** [comment]
Just to clarify, Fake WiFi has already been added back to the A13 build.
https://github.com/waydroid/android_vendor_waydroid/pull/44

**Comment by baptx** [comment]
@ayasa520 Was the A13 build already released or will it be released soon?

**Comment by ayasa520** [comment]
> [@ayasa520](https://github.com/ayasa520) Was the A13 build already released or will it be released soon?

https://sourceforge.net/projects/waydroid/

**Comment by baptx** [comment]
> > [@ayasa520](https://github.com/ayasa520) Was the A13 build already released or will it be released soon?
> 
> https://sourceforge.net/projects/waydroid/

I did not find "A13" so I guess you mean Android 13 (LineageOS 20) which has several builds.
Android 13 is already the version I installed a few days ago and `waydroid prop set persist.waydroid.fake_wifi package_name` was not enough to bypass the network check of an Android application.
So it looks like fake_wifi is not enough for some applications like someone mentioned here also:
https://github.com/waydroid/waydroid/discussions/1593
Fortunately the Frida solution worked. It seems needed for applications using the function getActiveNetworkInfo or its replacement getActiveNetwork.

**Comment by ayasa520** [comment]
So which app isn’t working? I need to know the exact package name. When the issue you mentioned was posted, that patch hadn’t even been merged into Waydroid yet.

**Comment by baptx** [comment]
It is a private application I used for work but I was able to reproduce the issue by using the Java code sample from https://stackoverflow.com/questions/57277759/getactivenetworkinfo-is-deprecated-in-api-29/57285217#57285217 in a simple Hello World application (https://www.tutorialspoint.com/android/android_hello_world_example.htm) and compiling it in Android Studio. It is also more reliable to reproduce the issue with our own code in case the code of a third-party Android application changes. I did not need the Application parameter in the isNetworkAvailable function because we can call directly getSystemService function.

The issue with the Waydroid fake_wifi seems to be that it only fakes NetworkCapabilities.TRANSPORT_WIFI and the application I was using only checked NetworkCapabilities.TRANSPORT_CELLULAR.
So a solution would be to add a fake_cellular property to Waydroid (better keep fake_wifi for WiFi only in case some weird apps will produce an error if we enable both cellular data and WiFi).
Faking cellular data is exactly what the Frida script https://codeshare.frida.re/@zionspike/bypass-wi-fi-check-on-android/ is doing (see the check "args == 0" at the end of the script, it means TRANSPORT_CELLULAR since it has the constant value 0 according to https://developer.android.com/reference/android/net/NetworkCapabilities#TRANSPORT_CELLULAR).

Another solution is using the Frida script but it currently only works if there is a delay of 1 second before the code we want to change, due to the Frida bug I mentioned in a previous comment.
I tested the delay with this code:
https://stackoverflow.com/questions/15874117/how-to-set-delay-in-android/15874176#15874176

**Comment by arkaroy14** [comment]
> It is a private application I used for work but I was able to reproduce the issue by using the Java code sample from https://stackoverflow.com/questions/57277759/getactivenetworkinfo-is-deprecated-in-api-29/57285217#57285217 in a simple Hello World application (https://www.tutorialspoint.com/android/android_hello_world_example.htm) and compiling it in Android Studio. It is also more reliable to reproduce the issue with our own code in case the code of a third-party Android application changes. I did not need the Application parameter in the isNetworkAvailable function because we can call directly getSystemService function.
> 
> The issue with the Waydroid fake_wifi seems to be that it only fakes NetworkCapabilities.TRANSPORT_WIFI and the application I was using only checked NetworkCapabilities.TRANSPORT_CELLULAR. So a solution would be to add a fake_cellular property to Waydroid (better keep fake_wifi for WiFi only in case some weird apps will produce an error if we enable both cellular data and WiFi). Faking cellular data is exactly what the Frida script https://codeshare.frida.re/@zionspike/bypass-wi-fi-check-on-android/ is doing (see the check "args == 0" at the end of the script, it means TRANSPORT_CELLULAR since it has the constant value 0 according to https://developer.android.com/reference/android/net/NetworkCapabilities#TRANSPORT_CELLULAR).
> 
> Another solution is using the Frida script but it currently only works if there is a delay of 1 second before the code we want to change, due to the Frida bug I mentioned in a previous comment. I tested the delay with this code: https://stackoverflow.com/questions/15874117/how-to-set-delay-in-android/15874176#15874176

I have solved the issue regarding many apps failed to detect network since they expect either Cellular or Wifi so I have made an xposed app works via lsposed u can check my repo for test app and hook app.
TRANSPORT_CELLULAR & TRANSPORT_WIFI both are bypassed in each cellular or wifi mode.
Frida scripts are not an optimized solution because the server and scripts need to be started every time. I alawys prefer implementing those frida logic into an Xposed module, where most apps can run directly under the LSPosed scope, unless its very secure and protected app.
[waydroid_network_spoof](https://github.com/arkaroy14/waydroid_network_spoof)

**Comment by baptx** [comment]
@arkaroy14 I think the downside of Xposed is that sometimes it is blocked by applications using root detection (LSPosed uses Magisk which can be detected with libraries like https://github.com/scottyab/rootbeer) so we need to bypass it.
Can you add the source code to your repository so we can build the APK ourselves?

**Comment by gustavokch** [comment]
# Restore WhatsApp Google Drive Backup via Frida WiFi Spoofing

WhatsApp requires an active WiFi connection before it will pull a backup from Google Drive. If you're on a rooted Android device without WiFi access — or migrating data and need to force a restore — you can hook the Android connectivity APIs using Frida to make WhatsApp believe it's on WiFi.

This guide covers everything from environment setup to a working attach-mode injection.

---

## Requirements

- Rooted Android device (tested on Pixel 5, Android 12+)
- ADB installed and working on your host machine
- Python 3.x on your host machine
- Frida 17.x installed on host: `pip install frida-tools`
- `frida-server` binary matching your Frida version, for `android-x86_64`
- The same phone number that owns the Google Drive backup

---

## Step 1 — Download frida-server

The version of `frida-server` on the device **must match** your host `frida-tools` version exactly. Check your version first:

```bash
frida --version
# e.g. 17.9.1
```

Then download the matching x86_64 binary:

```bash
wget https://github.com/frida/frida/releases/download/17.9.1/frida-server-17.9.1-android-x86_64.xz
unxz frida-server-17.9.1-android-x86_64.xz
mv frida-server-17.9.1-android-x86_64 frida-server
```

---

## Step 2 — Push frida-server to the device

```bash
adb push frida-server /data/local/tmp/frida-server
adb shell "su -c 'chmod 755 /data/local/tmp/frida-server'"
```

Start it in the background:

```bash
adb shell "su -c '/data/local/tmp/frida-server &'"
```

Confirm it's running:

```bash
adb shell "su -c 'ps | grep frida'"
```

You should see a `frida-server` process in the output. If you get a permission error, disable SELinux enforcement first:

```bash
adb shell "su -c 'setenforce 0'"
```

---

## Step 3 — Create the Frida script

Save the following as `spoof_wifi.js` on your host machine:

```javascript
Java.perform(function () {

    // --- Legacy NetworkInfo API ---
    var NetworkInfo = Java.use("android.net.NetworkInfo");

    NetworkInfo.getType.implementation = function () {
        return 1; // TYPE_WIFI
    };
    NetworkInfo.getTypeName.implementation = function () {
        return "WIFI";
    };
    NetworkInfo.isConnected.implementation = function () {
        return true;
    };
    NetworkInfo.isAvailable.implementation = function () {
        return true;
    };
    NetworkInfo.isConnectedOrConnecting.implementation = function () {
        return true;
    };
    NetworkInfo.getState.implementation = function () {
        var State = Java.use("android.net.NetworkInfo$State");
        return State.CONNECTED.value;
    };

    // --- Modern NetworkCapabilities API (API 21+) ---
    var NetworkCapabilities = Java.use("android.net.NetworkCapabilities");

    // TRANSPORT_WIFI = 1
    NetworkCapabilities.hasTransport.implementation = function (transport) {
        console.log("[*] hasTransport(" + transport + ")");
        if (transport === 1) return true;
        return this.hasTransport(transport);
    };

    NetworkCapabilities.hasCapability.implementation = function (cap) {
        console.log("[*] hasCapability(" + cap + ")");
        return true;
    };

    console.log("[+] WiFi spoof hooks active");
});
```

---

## Step 4 — Launch WhatsApp and attach Frida

**Do not use `-f` (spawn mode)** — it times out on newer Android versions due to SELinux and play integrity checks. Instead, launch WhatsApp manually on the device, navigate to the restore/setup screen, then attach from your host:

```bash
# Get the PID
adb shell "su -c 'pidof com.whatsapp'"

# Attach using the PID
frida -U -p <PID> -l spoof_wifi.js
```

Or attach by package name directly (no PID lookup needed):

```bash
frida -U -n com.whatsapp -l spoof_wifi.js
```

Once attached you should see:

```
[+] WiFi spoof hooks active
[*] hasTransport(1)
[*] hasCapability(12)
...
```

At this point tap the restore button in WhatsApp. The `[*]` log lines confirm which API calls are being intercepted.

---

## Step 5 — Complete the restore

With the hooks active, WhatsApp will pass its WiFi check and begin pulling the backup from Google Drive. Keep the Frida session open until the restore finishes — closing it will remove the hooks.

---

## Troubleshooting

**"Failed to spawn: unexpectedly timed out"**
Use attach mode (Step 4) instead of `-f`. Spawn mode is unreliable on production builds with Google Play integrity enabled.

**Still showing "Waiting for Wi-Fi" after attach**
Check the console output. If `hasTransport` and `hasCapability` lines are not appearing, WhatsApp may be running its check in a background service before you attached. Force-stop the app, reattach with `--await`, then open it:

```bash
adb shell "am force-stop com.whatsapp"
frida -U --await com.whatsapp -l spoof_wifi.js
# Then open WhatsApp manually on the device
```

**frida-server crashes on launch**
Version mismatch between host and device binaries is the most common cause. Re-check `frida --version` and re-download the exact matching server binary.

**SELinux blocking ptrace**
```bash
adb shell "su -c 'setenforce 0'"
adb shell getenforce  # should return: Permissive
```

---

## Notes

- Everything in this guide operates on your own device and your own backup data. frida-server requires root, so this will not work on a non-rooted device.
- Once the restore is complete, you can use WhatsApp's built-in **Chats > Export Chat** feature to move individual conversations to your iPhone, or use the official Move to iOS migration path if you haven't switched yet.
- Re-enable SELinux after you're done: `adb shell "su -c 'setenforce 1'"`

This was formatted by Claude, but it contains the exact steps I performed to get WhatsApp on Waydroid to restore media from a GDrive backup, which I couldn't do since switching to iOS.

<img width="789" height="368" alt="Image" src="https://github.com/user-attachments/assets/6382261b-cb54-4b76-a04d-ca782f4d381b" />

I hope this helps.
